### PR TITLE
[SPARK-37738][PYTHON] Support column type inputs for second arg of date manipulation functions

### PR DESCRIPTION
**Pull request by Daniel-Davies**
### What changes were proposed in this pull request?
See https://issues.apache.org/jira/browse/SPARK-37738

There seems to be a skew in the Scala Spark API & PySpark API; namely, date_add/ date_sub/ add_months take an 'int' type for the 'days' parameter in PySpark, but can accept a column or an integer in Scala.

This PR makes both types available to the 'days' parameter in PySpark.

### Why are the changes needed?
Users should see a consistent API across Python & Scala side processing.

### Does this PR introduce _any_ user-facing change?
Yes- additive only.

>>> df = spark.createDataFrame([('2015-04-08', 2,)], ['dt', 'add'])
>>> df.select(date_add(df.dt, df.add).alias('next_date')).collect()
[Row(next_date=datetime.date(2015, 4, 10))]

### How was this patch tested?
3 new unit tests

### CODE REVIEW

File: python/pyspark/sql/functions.py
Comment (zero323) [inline python/pyspark/sql/functions.py:2210]: We don't really need overloads here ‒ these are useful when there are some relationships between arguments and / or return type.

**Comment by zero323** [comment]
Also, it might be a good idea to fix that

No user is associated with the committer email

File: python/pyspark/sql/functions.py
Comment (Daniel-Davies) [reply python/pyspark/sql/functions.py:2210]: ack- will remove, thank you! I thought it would be a good idea for consistency with the other functions, but you're right in that date_add (+ others I've modified) will output the same thing regardless

**Comment by Daniel-Davies** [comment]
> Also, it might be a good idea to fix that
> 
> No user is associated with the committer email

Oops 😅 working cross github accounts atm, thank you for the flag!

File: python/pyspark/sql/functions.py
Comment (zero323) [inline python/pyspark/sql/functions.py:2231]: It might be better to wrap `days` with `lit` outside the block:


```python
days = days if isinstance(days, Column) else lit(days)

return sc._jvm.functions.date_add(_to_java_column(start), _to_java_column(days))
```

Same for the other ones.

Comment (Daniel-Davies) [reply python/pyspark/sql/functions.py:2221]: Sure- I like this better ^^ thanks for the suggestion!

Comment (Daniel-Davies) [reply python/pyspark/sql/functions.py:2221]: I assume that it will still work if I passed a string-name rather than a spark column, right?

Comment (zero323) [reply python/pyspark/sql/functions.py:2221]: Alternatively, with `_create_column_from_literal`

```
_create_column_from_literal
days = days._jc if isinstance(days, Column) else _create_column_from_literal(days)

return sc._jvm.functions.date_add(_to_java_column(start), days)
```

Comment (Daniel-Davies) [reply python/pyspark/sql/functions.py]: This also wouldn't allow me to write:

df = df.withColumn("new_date", F.date_add("old_date", "int_column"))

Or am I mistaken?

Comment (Daniel-Davies) [reply python/pyspark/sql/functions.py]: ` days = lit(days) if isinstance(days, int) else days

    return Column(
        sc._jvm.functions.date_add(
            _to_java_column(start), _to_java_column(days)
        )
    )`

Comment (Daniel-Davies) [reply python/pyspark/sql/functions.py]: This works ^^ but it's not optimal if you want to avoid the check on the int; is there a benefit you see to running the isinstance on the Column class specifically?

Comment (zero323) [reply python/pyspark/sql/functions.py]: You're right. One way or another we have to check on `int`, so maybe the initial approach was OK. Sorry for the confusion.

Comment (Daniel-Davies) [reply python/pyspark/sql/functions.py]: No worries at all! I've amended the code to the above ^^ so keeping a refactor of the 'days' variable- thank you for suggesting this.

**Comment by zero323** [comment]
cc @HyukjinKwon.

File: python/pyspark/sql/tests/test_functions.py
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py:290]: you can import this on the top, and let's add a comment like # SPARK-XXXXX: blah blah (see also https://spark.apache.org/contributing.html)

**Review by HyukjinKwon — APPROVED** [review]
The API change LGTM. @zero323 is touching some codes around here so would be better to have his signoff here.

File: python/pyspark/sql/tests/test_functions.py
Comment (Daniel-Davies) [reply python/pyspark/sql/tests/test_functions.py]: Thank you for the review @HyukjinKwon - have amended as per your comments.

File: python/pyspark/sql/tests/test_functions.py
Comment (zero323) [inline python/pyspark/sql/tests/test_functions.py:309]: This seems to be obsolete.

**Comment by Daniel-Davies** [comment]
Great, thanks both- @zero323 let me know if there's anything else.

I also have one other question; when is the Column class used vs ColumnOrName? E.g., why are both types used in the following function?

https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L3695

Would it be worth converting some of the existing Column type inputs to use ColumnOrName?

**Comment by HyukjinKwon** [comment]
That's related to this https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L60-L64. It's because of legacy behaviours. Yes we should convert all to `ColumnOrName`. There are some special cases where `Column` and `str` mean different things (e.g., https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L4795-L4796). In such cases, we should explicitly document that `str` here does not mean `Column` names.

File: python/pyspark/sql/tests/test_functions.py
Comment (zero323) [inline python/pyspark/sql/tests/test_functions.py:302]: We could skip multiple jobs here (and add tests for column name as well):

```
self.assertEqual(all(df.select(
    date_add(df.date, df.add) == datetime.date(2021, 12, 29),
    date_add(df.date, "add") == datetime.date(2021, 12, 29),
    date_add(df.date, 3) ==  datetime.date(2021, 12, 30)
).first()))
```

It is not much, but we have a lot of functions and these quickly add up.

File: python/pyspark/sql/tests/test_functions.py
Comment (zero323) [inline python/pyspark/sql/tests/test_functions.py:295]: How about we add schema here

```python
df = self.spark.createDataFrame([Row(date=dt, add=2)], "date date, add integer")
```

**Comment by zero323** [comment]
Other than these, LGTM.

**Comment by Daniel-Davies** [comment]
@HyukjinKwon - All makes sense to me; I'd be happy to write this in a ticket and make a PR to correct some of the input parameters over the next few days.

Thank you for the review @zero323 - I've amended as per your comments!

File: python/pyspark/sql/tests/test_functions.py
Comment (Daniel-Davies) [reply python/pyspark/sql/tests/test_functions.py]: I didn't know you could specify schema input like this- have amended; thank you!

File: python/pyspark/sql/tests/test_functions.py
Comment (Daniel-Davies) [reply python/pyspark/sql/tests/test_functions.py]: Completely agree, and definitely much easier to read

**Comment by zero323** [comment]
> Thank you for the review @zero323 - I've amended as per your comments!

Thanks. `[:]` shouldn't be necessary. `Row` is just fancy tuple and should work just fine in `all`.

**Comment by Daniel-Davies** [comment]
> > Thank you for the review @zero323 - I've amended as per your comments!
> 
> Thanks. `[:]` shouldn't be necessary. `Row` is just fancy tuple and should work just fine in `all`.

Ah so it does, apologies; think this was to fix an unrelated issue

**Comment by zero323** [comment]
If you check the failing tests you'll see that we have missing import for `IntegerType`. Alternatively you can just use string name (`"integer"`) :)

Also, we should change title to something more descriptive. Maybe "Support column as a second argument for date adding functions"?

**Comment by Daniel-Davies** [comment]
ah ouch, didn't see that one coming- my apologies

**Comment by zero323** [comment]
No worries. Sadly, linters don't catch these. 

As of this:

> All makes sense to me; I'd be happy to write this in a ticket and make a PR to correct some of the input parameters over the next few days.

‒ we don't seem to have many of these left. `least` (which is matter of fixing the annotation, as code already supports it) is one.  The other one is `when` which is somewhat disputable; value and otherwise can take plain literals (including strings)  which are interpreted as-is

```python
>>> spark.range(1).select(
...     when(lit(False), "foo").otherwise("bar"),
...     when(lit(True), "foo").otherwise("bar")
... ).show()
##  +-------------------------------------+------------------------------------+
## |CASE WHEN false THEN foo ELSE bar END|CASE WHEN true THEN foo ELSE bar END|
## +-------------------------------------+------------------------------------+
## |                                  bar|                                 foo| 
## +-------------------------------------+------------------------------------+

```

so I am not sure if we want to change anything there. This

```python
when(col("baz"), "foo").otherwise("bar")
```

is explicit, while

```python
when("baz", "foo").otherwise("bar")
```

might be confusing, suggesting that all strings have the same meaning, which is not the case.

**Comment by Daniel-Davies** [comment]
Yes, the ones I saw were:

- F.when: Completely agree, I think we should leave this one alone
- F.overlay: Annotation only
- F.least: Annotation only
- F.slice: Needs a conversion
- F.array_repeat: Needs a conversion
- F.percentile_approx: I don't think this should take a column name either way though

So really there are only two annotation changes & two column conversions needed from what I can see. Happy to update these, or if you don't think they're worth pursuing/ have reasons for being this way, we may want to leave them alone.

**Comment by zero323** [comment]
> F.percentile_approx: I don't think this should take a column name either way though

Yeah, this one takes `Columns` only to reduce number of overloaded signatures on Scala side.

**Review by zero323 — APPROVED** [review]

**Comment by zero323** [comment]
Merged to master.

**Comment by Daniel-Davies** [comment]
Thank you both, appreciate your guidance in this PR!

**Pull request closed without merging**
